bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/launch/EquinoxFactory.java

```diff
@@ -23,15 +29,40 @@
  * The framework factory implementation for the Equinox framework.
  * @since 3.5
  */
-public class EquinoxFactory implements FrameworkFactory, ConnectFrameworkFactory {
+public class EquinoxFactory implements FrameworkFactory, ConnectFrameworkFactory, FrameworkUtilHelper {
+
+	static final Set<FrameworkUtilHelper> connectHelpers = ConcurrentHashMap.newKeySet();
 
 	@Override
 	public Framework newFramework(Map<String, String> configuration) {
-		return newFramework(configuration, null);
+		return new Equinox(configuration);
 	}
 
 	@Override
-	public Framework newFramework(Map<String, String> configuration, ModuleConnector moduleConnector) {
-		return new Equinox(configuration, moduleConnector);
+	public ConnectFramework newFramework(Map<String, String> configuration, ModuleConnector moduleConnector) {
```

Comment:
A complication with this approach is the two views of the system bundle (`Framework`) that Equinox currently has.  We have the external view from the launcher `Equinox` class that implements `Framework` as well as the internal view with the `org.eclipse.osgi.internal.framework.EquinoxBundle.SystemBundle` class.  The `Equinox` class ends up delegating to the internal representation of the system bundle class `org.eclipse.osgi.internal.framework.EquinoxBundle.SystemBundle` after it has created the `EquinoxContainer`.

This fact allowed you to provide an alternative view (`ConnectFramework`) externally so that external launchers can add their helpers.  This is good because internally bundles can cast `Bundle` with ID=0 to `Framework` but they would not be able to cast to `ConnectFramework` and therefore never be able to add helpers internally (something I wanted to block).  But other OSGi framework implementations may not implement the system bundle this way.  They may actually have the internal system bundle representation be the exact same object returned from the `FrameworkFactory`.

An alternative approach could be to have an interface which `ModuleConnector` implementations can also implement that has a `getFrameworkUtilHelper()` method.  When a framework instance is initialized with a `ModuleConnector` that provides a helper then the framework can get that helper registered in some implementation specific way.

This also allows us to avoid having to manage the helpers with add/remove methods which can lead to leaky bugs.  Instead when a framework is stopped it will need to clean up the `ModuleConnector` helper from its internal store so it is no longer called by the `FrameworkUtil` class.  If that stopped framework instance is initialized again then the helper from the `ModuleConnector` would be applied again to the frameworks helper automatically.

I propose you do not prototype any changes to the `org.osgi` package.  Instead you could add an interface to the `org.eclipse.osgi.launch` package that holds the `Optional<FrameworkUtilHelper> getFrameworkUtilHelper()` method which we would like to eventually get added (as a default method returning empty) to the `ModuleConnector` interface in a future OSGi specification.

Before we merge anything like that I would like to get some agreement on this direction in the OSGi Spec working group to avoid adding API to Equinox that may be drastically different from what would eventually be spec'ed in OSGi.

Reply:
> An alternative approach could be to have an interface which `ModuleConnector` implementations can also implement that has a `getFrameworkUtilHelper()` method. When a framework instance is initialized with a `ModuleConnector` that provides a helper then the framework can get that helper registered in some implementation specific way.

Thinking more about this, maybe one can simply say, if `ModuleConnector` also implements `FrameworkUtilHelper` that would be used? I think this would be most "unintrusive" and even only need a spec adjustment instead of changing any code or interfaces, I'll update the PR accordingly so we can see how it behaves...

> Before we merge anything like that I would like to get some agreement on this direction in the OSGi Spec working group to avoid adding API to Equinox that may be drastically different from what would eventually be spec'ed in OSGi.

If I remember correctly the next two calls are canceled, maybe we can still use the time to came up with something we are comfortable with so its much easier to discuss in the next spec call then, if we can agree on a method we might even can prepare a PR to change the spec accordingly as a base for the discussion, I just never did this before so no idea how complicate it would be to do so...

Reply:
> Thinking more about this, maybe one can simply say, if `ModuleConnector` also implements `FrameworkUtilHelper` that would be used? I think this would be most "unintrusive" and even only need a spec adjustment instead of changing any code or interfaces, I'll update the PR accordingly so we can see how it behaves...

Sure, that would be the least amount of API change and something we could prototype without breaking any signature tests.  But it is hard to adequately document such implicit contracts without a visible API signature change.  I think I prefer the more explicit API change that documents the behavior clearly.

Reply:
> I just never did this before so no idea how complicate it would be to do so...

Specs move slow, so I would be prepared to have an interim solution that is Equinox specific for a while.  I just want to ensure we are happy to support that Equinox specific API for the long run.

Reply:
> But it is hard to adequately document such implicit contracts without a visible API signature change. I think I prefer the more explicit API change that documents the behavior clearly.

What do you think about using the (implicit) solution for this PR (to not break signatures) and suggest to the Spec that `ModuleConnector` simply extends `FrameworkUtilHelper` to make it explicit? That sounds like the most simple solution as it also removes any need for specifying that a (whatever) getter always return the same instance and so on.

Reply:
> What do you think about using the (implicit) solution for this PR (to not break signatures) and suggest to the Spec that `ModuleConnector` simply extends `FrameworkUtilHelper` to make it explicit? That sounds like the most simple solution as it also removes any need for specifying that a (whatever) getter always return the same instance and so on.

Seems like a reasonable starting point.  It also reduces the risk to the Equinox API.  Unless/until we document this somehow it remains an internal dependency.  Tycho can certainly depend on the internal, but we would be free to remove the internal support until we documented it somehow as a proper external.  I worry about such things because no API tooling can detect changes to such implicit support.  Hence I would prefer it remain something internal that we can play around with until there is a spec'ed way to do it.

At the spec level we would have to determine if extending `FrameworkUtilHelper` can happen in a non-major version (breaking) way by adding the required default implementation directly to `ModuleConnector`.  `ModuleConnector` is a consumer type so we need to add methods to it carefully.

But thinking about that, it would then force the framework implementation to always delegate to the supplied `ModuleConnector` even though it never provides an overriding `org.osgi.framework.connect.FrameworkUtilHelper.getBundle(Class<?>)` method.  That may be a contentious point to discuss at the spec level because it is not something a `ModuleConnector` can opt into with a mix-in interface like we would do at first in Equinox with optionally allowing a `ModuleConnector` to implement `FrameworkUtilHelper`.  Ultimately that may result in OSGi deciding to just support the mix-in approach.  But then they would at least add that to the javadoc of `ConnectFrameworkFactory` and `ModuleConnector`

Reply:
> Hence I would prefer it remain something internal that we can play around with until there is a spec'ed way to do it.

That would be fine for me, I already got the official way working for 80% of the cases and switching to felix or some other osgi impl seems not really an option here.

> At the spec level we would have to determine if extending `FrameworkUtilHelper` can happen in a non-major version (breaking) way by adding the required default implementation directly to `ModuleConnector`. `ModuleConnector` is a consumer type so we need to add methods to it carefully.

I more though about `ModuleConnector` extends `FrameworkUtilHelper` so `FrameworkUtilHelper` remains unchanged or do I understand wrong?

> But thinking about that, it would then force the framework implementation to always delegate to the supplied `ModuleConnector` even though it never provides an overriding

I think this is acceptable and assume most (if not all) implementations will migrate to the `ModuleConnector` approach as the SPI one is really hard to get right... But if all this is a concern, I think we can also add a method to `ConnectFrameworkFactory` that accepts a `ModuleConnector` and a `FrameworkUtilHelper` that should then only be a provider type change ...